### PR TITLE
Fix events in hooks triggered by a move

### LINKED ISSUE #482: events.js does not handle any events that are further generated in the code that it runs
For example:

```js
const game = {
  moves: {
    endTurnMove: (G, ctx) => {
      ctx.events.endTurn();
    },
  },
  turn: {
    onEnd: (G, ctx) => {
      ctx.events.endPhase();
    },
  },
  phases: {
    A: { start: true, next: 'B' },
    B: {},
  },
};

test('this test passes', () => {
  const client = Client({ game });
  expect(client.getState().ctx.phase).toBe('A');

  client.events.endTurn();

  expect(client.getState().ctx.currentPlayer).toBe('1');
  expect(client.getState().ctx.phase).toBe('B');
});

test('this test fails', () => {
  const client = Client({ game });
  expect(client.getState().ctx.phase).toBe('A');

  client.moves.endTurnMove();

  expect(client.getState().ctx.currentPlayer).toBe('1');
  expect(client.getState().ctx.phase).toBe('B');    // phase is still A
});
```

This is due to the fact that events are processed only [once](https://github.com/nicolodavis/boardgame.io/blob/master/src/core/reducer.js#L191).

### LINKED ISSUE #487: setActivePlayers function in onBegin hooks only work once
```js
{
  turn: {
    onBegin: (G, { numPlayers, currentPlayer, events }) => {
      const value = {};
      for (let i = 0; i < numPlayers; i++) {
        value[i] = currentPlayer === String(i) ? 'main' : 'others';
      }
      events.setActivePlayers({ value });
      return G;
    }
  }
};
```

in the next turn `ctx.activePlayers` will be null

### PR DESCRIPTION
**Pull request by delucis**
Closes #482
Closes #487

The `update` method of the `Events` class currently caches the length of the event dispatch queue before processing it, avoiding processing of any new events pushed during processing.

https://github.com/nicolodavis/boardgame.io/blob/e20aca433f758ab66eaabbd077f61cebc25a9575/src/core/events.js#L42-L44

This seems sensible as it limits the scope of `update ` and presumably guards against infinite loops. However, this also causes the bug that events triggered _after_ a move, i.e. in hooks, are never processed and discarded.

This PR adds a test for the expected behaviour and simply removes the caching of the queue length, meaning now those new events are handled correctly. All tests are passing, but if you think there’s a risk making this change, let me know.

---

**Edit:** Just noticed [the commit that added the length caching](https://github.com/nicolodavis/boardgame.io/commit/39c00e41690f0d40c4b3f5ac47213f023ff53f6d) was labelled “remove infinite loop”, but the infinite loop tests seem to be passing, so maybe something else is now preventing infinite loops?

### CODE REVIEW

File: src/core/events.js [hunk 1]
```diff
@@ -40,8 +40,7 @@ export class Events {
    * @param {object} state - The state object { G, ctx }.
    */
   update(state) {
-    const length = this.dispatch.length;
-    for (let i = 0; i < length; i++) {
+    for (let i = 0; i < this.dispatch.length; i++) {
```
Comment (nicolodavis) [inline src/core/events.js]: Let's limit this to 100 iterations to catch infinite loops. We can probably log a message if this loop reaches the 100th iteration.

Comment (nicolodavis) [reply src/core/events.js]: It's possible that some other code in `flow.js` is now catching infinite loops.

Comment (delucis) [reply src/core/events.js]: I’ve tried adding two more tests to see if I can trigger an infinite loop, but they seem to be caught elsewhere. Given these tests, maybe it’s safe to assume that an iteration limit is unnecessary?

Comment (nicolodavis) [reply src/core/events.js]: OK, sounds good to me.

File: src/core/flow.test.js [hunk 2]
```diff
@@ -932,6 +932,40 @@ describe('infinite loops', () => {
...
+
+  test('loop 4', () => {
+    const game = {
+      moves: {
+        endTurn: (G, ctx) => {
+          ctx.events.endTurn();
+        },
+      },
+      turn: {
+        endif: () => true,
```
Comment (nicolodavis) [inline src/core/flow.test.js]: This should be `endIf`.

**Comment by nicolodavis** [comment]
Found another related bug #526.

**Comment by delucis** [comment]
Ok, this should be good to go. Adding an iteration limit is an easy fix if we do find some infinite loops or fixing other event-related bugs reintroduces the potential for them.

**Pull request merged**